### DISCUSSION TITLE
Make Quill instances reconfigurable

### DISCUSSION THREAD

**Discussion by singintime** [slab/quill#4397] [Ideas]
## Motivation

It's not currently possible to reconfigure an already existing instance of Quill.

The only way to change the editor's configuration is to create an entirely new instance and replace the old one. This operation has several undesirable side effects:
- It may cause performance issues and memory leaks, caused by dangling subscriptions to event listeners,
- Replacing a focused DOM element triggers a focus event, leading to potential collisions with other events such as `aria-live` updates.

The latter case is exactly what happens in [Miro](https://miro.com). The objects in Miro boards use the Quill editor to provide rich text editing capabilities. Every object type has a different configuration of the editor, so every time a new object type is selected the (already focused) Quill's root on the page is replaced. The triggered focus event silences a very important `aria-live` update that provides detailed information about the selected object on the board.

[Recent updates in Google Chrome](https://chromium.googlesource.com/chromium/src/+/40942fad2bc11700eda3b96275b4b3b8d81ac18a) made this issue even worse. Focus events could once be silenced by setting the `aria-hidden` attribute on the Quill editor, but now they can't anymore. On top of that, the latest versions of Apple VoiceOver don't allow live region updates to interrupt focus announcements. This unfortunate combination entirely breaks the user experience of Chrome + VoiceOver screen reader users.

## Proposed changes

Making Quill instances reconfigurable, by:
- Replacing all lambdas with named functions in event subscriptions,
- Adding a `detach` method to all Quill components and modules that register event subscriptions,
- Using the above-mentioned detach methods to unsubscribe from all events before reconfiguring the instance,
- Providing a `configure` method in the main Quill class to apply a new configuration to the editor.

**Comment by singintime** [comment]
After a more accurate search, it looks like there's a similar issue already open: https://github.com/slab/quill/issues/4234

**Comment by bmakuh** [comment]
I think that older issue is somewhat related, but I really like this approach of providing the ability to tear down an instance via `detach()` methods. This seems a lot better than just mutating options/properties on an already initialized instance.
